oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBAggregationQueryDAO.java

```diff
@@ -85,24 +60,45 @@ protected void apply(MeasureQuery query) {
             throw new IOException("field spec is not registered");
         }
 
+        if (schema.getTopNSpec() == null) {
+            throw new IOException("TopN spec is registered");
+        }
+
+        TopNQueryResponse resp = null;
+        if (condition.getOrder() == Order.DES) {
+            resp = topN(schema, timestampRange, condition.getTopN(), additionalConditions);
+        } else {
+            resp = bottomN(schema, timestampRange, condition.getTopN(), additionalConditions);
+        }
```

Comment:
`additionalConditions` may contain `service_id` or not. Do we need two TopN annotations for two rules?

Reply:
I think w/ and w/o service ID mean different pre-ordered at BanyanDB side for pre-calculation.

Reply:
If we need two annotations at the class level, we need another annotation(syntactic sugar of TopN)

Reply:
I think we should always add complete groupByTags to the annotation in order to make filter works.

BanyanDB client can restore those groupBy tags, thus we only need one TopN processor in the database.

Reply:
I mean there are two kinds of topN.
1. Where **service_id = xxx** and time_range in [start, end] Group by entityID
2. Where and time_range in [start, end] Group by entityID

Check the query condition protocol. 
https://github.com/apache/skywalking-query-protocol/blob/450328c6445d4dc50a868a0b797a14755f1d82cc/metrics-v2.graphqls#L74-L75

Reply:
> I mean there are two kinds of topN.
> 
> 1. Where **service_id = xxx** and time_range in [start, end] Group by entityID
> 2. Where and time_range in [start, end] Group by entityID
> 
> Check the query condition protocol. https://github.com/apache/skywalking-query-protocol/blob/450328c6445d4dc50a868a0b797a14755f1d82cc/metrics-v2.graphqls#L74-L75

BanyanDB supports post-aggregation as is hinted in the protocol

https://github.com/apache/skywalking-banyandb/blob/main/api/proto/banyandb/measure/v1/topn.proto#L62

I suppose it can serve the first case

Reply:
I think server side should treat topN more complex. Such as, we have conditions(service id) as filter, put pre aggregated list according to service candidate?

Anyway, we need a solution for this. Otherwise, this is a bug. In a distributed system, some services are slower than others.

Reply:
> > Let's say there are 200 services, each service has 50k endpoints, we need both service-based topN endpoints and global level topN endpoints. How do we make this work? If we only set the rule as top 1000, which are all belonging to several services. This means we can't have topN for other services.
> 
> > 
> 
> > @lujiajing1126 @hanahmily What are your suggestions for this real case?
> 
> 
> 
> Yes. It is possible. Even two TopN(s) or more cannot help in this case.

From the server’s perspective, we can set a global endpoint topn process and several sub-endpoints top with a service id to support this scenario. @lujiajing1126 What’s the gap here?

Reply:
How to define that? The service ID is not predictable. We just know service ID should be used as `group by` condition.

Reply:
@lujiajing1126 I did a discussion with @hanahmily 
I think the server side is clear, and the thing is you seem mixed query groupBy condition and topN pre-aggregation GroupBy condition.

In the current case, we should have two TopN definition per metric
1. Global TopN 
2. Service-ID oriented groupBy TopN

Notice, the entityID should never be involved pre-aggregation TopN. That ID is the groupBy condition in the query stage only about how to merge data cross bucket. But pre-aggregation is bucket-based.

I mentioned to @hanahmily, in theory <1> could be built through all <2>'s TopN by costing some extra CPU. Considering <1> oriented query is supported but not widely used, we could enhance the server side to build <1> in the query stage only. You could discuss with Gao to see whether we could do this.

Reply:
So about the annotation, we could keep  `@BanyanDB.TopNAggregation(groupByTagNames = {"SERVICE_ID"})` for all OAL and MAL function, and let `groupByTagNames` accepts the column not existing. Then we even dont' need to update OALRuntime, but keep the annotation on the function(OAL and MAL) only.
WDYT?